components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,246 @@
+  if (_rippleColors == nil) {
+    _rippleColors = [NSMutableDictionary dictionary];
+    UIColor *selectionColor = RippleSelectedColor();
+    _rippleColors[@(MDCRippleStateNormal)] = [UIColor colorWithWhite:0 alpha:kDefaultRippleAlpha];
+    _rippleColors[@(MDCRippleStateHighlighted)] = [UIColor colorWithWhite:0
+                                                                    alpha:kDefaultRippleAlpha];
+    _rippleColors[@(MDCRippleStateSelected)] =
+        [selectionColor colorWithAlphaComponent:kDefaultRippleSelectedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateHighlighted)] =
+        [selectionColor colorWithAlphaComponent:kDefaultRippleAlpha];
+    _rippleColors[@(MDCRippleStateDragged)] = [UIColor colorWithWhite:0
+                                                                alpha:kDefaultRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateDragged | MDCRippleStateHighlighted)] =
+        [UIColor colorWithWhite:0 alpha:kDefaultRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
+        [selectionColor colorWithAlphaComponent:kDefaultRippleDraggedAlpha];
+  }
+  self.userInteractionEnabled = YES;
+}
+
+- (UIColor *)rippleColorForState:(MDCRippleState)state {
+  UIColor *rippleColor = _rippleColors[@(state)];
+  if (rippleColor == nil && (state & MDCRippleStateDragged) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateDragged)];
+  } else if (rippleColor == nil && (state & MDCRippleStateSelected) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateSelected)];
+  }
+
+  if (rippleColor == nil) {
+    rippleColor = _rippleColors[@(MDCRippleStateNormal)];
+  }
+  return rippleColor;
+}
+
+- (void)updateRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setRippleColor:rippleColor];
+}
+
+- (void)updateActiveRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setActiveRippleColor:rippleColor];
+}
+
+- (void)setRippleColor:(UIColor *)rippleColor forState:(MDCRippleState)state {
+  _rippleColors[@(state)] = rippleColor;
+
+  [self updateRippleColor];
+}
+
+- (MDCRippleState)state {
+  NSInteger state = 0;
+  if (self.selected) {
+    state |= MDCRippleStateSelected;
+  }
+  if (self.rippleHighlighted) {
+    state |= MDCRippleStateHighlighted;
+  }
+  if (self.dragged) {
+    state |= MDCRippleStateDragged;
+  }
+  return state;
+}
+
+- (void)setSelected:(BOOL)selected {
+  if (!self.allowsSelection) {
+    // If we disallow selection we don't want to apply any visual or state changes for selection.
+    return;
+  }
+  if (_tapWentOutsideOfBounds) {
+    // If the tap goes outside of bounds when a selection state change is triggered, we want to
+    // return early and not issue the selection state change as guidelines dictate that if a tap is
+    // let go outside the bounds, it should not trigger an action like issuing
+    // a selection/deselection.
+    return;
```

Comment:
Out of curiousity, what happens if multiple taps happen on this view? E.g. first tap down, drag out, second tap down, first tap release, second tap release.

Reply:
That's a good use case. It was acting pretty well with a slight caveat that I fixed.

When the first tap went outside of bounds the ripple faded out, and i expected the second tap to not issue a new ripple as the first tap is being held. That was the case aside from a small millisecond where an overlay appeared then instantly disappeared, and that was caused due to touchesMoved fading out the ripple again in line 238. To solve this unexpected glitch I made sure that the initialization of the fade in/out values that is set in `pointInside` is only initialized if we hadn't yet received a touch. It now works as expected, and the second tap doesn't invoke a new ripple and the first tap when it goes back into the bounds will fade the ripple back in.